### PR TITLE
fileserver: fix breadcrumbs for directories with % percent sign in the name

### LINKED ISSUE #5143: fileserver: Breadcrumbs are broken if directory name contains percent sign %
Breadcrumbs are generated without text if directory name contains a percent sign.
The bug is present in both

```
> caddy version
v2.6.1 h1:EDqo59TyYWhXQnfde93Mmv4FJfYe00dO60zMiEt+pzo=
```

and also the latest master 6bad878a22e048762262d6fabe2144cefaf4ca81

To replicate:  
Take this file tree 

```
> tree
.
└── dir with % in the name
    └── another
        └── with %
            └── sub
```

and run caddy with `caddy file-server --browse`

it creates this HTML:

```html
<header>
	<h1>
		<a href="../../../../">/</a><a href="../../../"></a>/<a href="../../">another</a>/<a href="../"></a>/<a href="">sub</a>/
	</h1>
</header>
```

Note the missing link text for directories with `%` signs in their name.

A test case to replicate the issue:

```golang
// put this in
// modules/caddyhttp/fileserver/browsetplcontext_test.go:21

func TestBreadcrumbs(t *testing.T) {
	testdata := []struct {
		path     string
		expected []crumb
	}{
		{"folder/with % sign/sub", []crumb{
			{Link: "../../", Text: "folder"},
			{Link: "../", Text: "with % sign"}, // .Text == "" here
			{Link: "", Text: "sub"},
		}},
```

### PR DESCRIPTION
**Pull request by abdusco**
Closes #5143

### CODE REVIEW

File: modules/caddyhttp/fileserver/browsetplcontext.go [hunk 1]
```diff
@@ -157,14 +157,22 @@ func (l browseTemplateContext) Breadcrumbs() []crumb {
...
 	return result
 }
 
+func unescapeSlash(s string) string {
+	// url.PathUnescape can't handle all uses of % symbol
+	// because it is used to encode special chars in URLs.
+	// "a % b" is valid folder name, but an invalid percent-encoding.
+	s = strings.ReplaceAll(s, "%2F", "/")
+	return strings.ReplaceAll(s, "%2f", "/")
+}
```
Comment (abdusco) [inline modules/caddyhttp/fileserver/browsetplcontext.go:174]: would the use of `strings.Replacer` be appropriate here?

File: modules/caddyhttp/fileserver/browsetplcontext.go [hunk 2]
```diff
@@ -157,14 +157,22 @@ func (l browseTemplateContext) Breadcrumbs() []crumb {
...
+
```
Comment (abdusco) [inline modules/caddyhttp/fileserver/browsetplcontext.go:175]: would the use of `strings.Replacer` be appropriate here?

**Review by mohammed90 — COMMENTED** [review]
Great fix! I was worried it might break other special characters, but my testing shows no breakage for them (e.g. `/with @ in dirname`). You can probably add tests for them.

You can use the same function to fix line lines 95-97. I've tested it locally and it works. Once that is done, this PR is good to go 👍🏼

**Comment by abdusco** [comment]
> You can use the same function to fix line lines 95-97. I've tested it locally and it works. Once that is done, this PR is good to go 👍🏼

@mohammed90 I'm writing tests for it, and I am not sure what the expected behavior is:

in `percent-encoded slash` case, I'd expect to receive `with/slash` as the name, not `slash`

```golang
func TestFileServer_directoryListing(t *testing.T) {
	tests := []struct {
		name         string
		urlPath      string
		expectedName string
	}{
		{
			name:         "no-encoding",
			urlPath:      "/with%percent",
			expectedName: "with%percent",
		},
		{
			name:         "encoded space",
			urlPath:      "/with%20space",
			expectedName: "with%20space",
		},
		{
			name:         "encoded slash %2F in dir name",
			urlPath:      "/with%2Fslash",
			expectedName: "with/slash",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			fsrv := FileServer{}
			bCtx := fsrv.directoryListing(
				context.Background(),
				nil,
				false,
				"/path/to/root",
				tt.urlPath,
				caddy.NewEmptyReplacer(),
			)
			actual := bCtx.Name
			if actual != tt.expectedName {
				t.Errorf("got %+v but expected %+v", actual, tt.expectedName)
			}
		})
	}
}
```

therefore I think changing this is also necessary:

```diff
-       name, _ := url.PathUnescape(urlPath)
+       name := unescapeSlash(path.Base(urlPath))
        return browseTemplateContext{
-               Name:     path.Base(name),
+               Name:     name,
```

or drop unescaping slashes altogether, because why slash get to be treated differently? Why don't encoded spaces also get decoded?

**Comment by abdusco** [comment]
@francislavoie It's not clear to me why we need to decode percent-encoded chars (specifically slashes %2f)

https://github.com/caddyserver/caddy/blob/master/modules/caddyhttp/fileserver/browsetplcontext.go#L157-L160

https://github.com/caddyserver/caddy/blob/6bad878a22e048762262d6fabe2144cefaf4ca81/modules/caddyhttp/fileserver/browsetplcontext.go#L95

I think we should drop decoding percent-encoded chars altogether.

**Comment by abdusco** [comment]
OK, I found why `url.PathUnescape` was added in the first place. I remember experiencing this issue as well: 

https://github.com/caddyserver/caddy/pull/4410

 

I think we can use a more polite approach and have:

```
decoded, err := url.PathUnescape(path)
if err != nil {
  decoded = path
}
```
@mohammed90 what do you think?

**Comment by mohammed90** [comment]
> OK, I found why `url.PathUnescape` was added in the first place. I remember experiencing this issue as well:
> 
> #4410
> 
> I think we can use a more polite approach and have:
> 
> ```
> decoded, err := url.PathUnescape(path)
> if err != nil {
>   decoded = path
> }
> ```
> 
> instead of removing url decoding, which would bring back the old issue.
> 
> @mohammed90 what do you think?

I think that's very reasonable. Sorry, I don't have much time on hand to validate the displayed text. If you don't mind, have a few directories/files with various non-ASCII characters to double-check.

**Comment by mholt** [comment]
Thanks for contributing this! Sorry I'm just now getting to it.

It seems to me that if PathUnescape is failing because of `%` not being escaped, the path has already been unescaped. :thinking:

While I think this is probably a fine solution, a more elegant one would be to only escape the path once.

I'm having trouble testing locally this since the code that added PathUnescape() was to fix when directories contain "/" in their names, which is impossible on Linux. And I run Linux/Mac, not Windows.

Would it be too tedious to find a solution that doesn't involve special-casing things, and instead we just Unescape once?

**Pull request closed without merging**